torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -379,7 +377,15 @@ def _init_param_attributes(self, p: Parameter) -> None:
         assert hasattr(p, "_is_sharded") and hasattr(
             p, "_orig_size"
         ), "Parameters should have been sharded during construction."
+        # If _local_shard has been set in the first lazy init and
+        # current parameter is pointed to _local_shard, no need to
+        # set the _local_shard again.
         if hasattr(p, "_local_shard"):
+            assert p.data_ptr() == p._local_shard.data_ptr(), (  # type: ignore[attr-defined]
+                "Parameter storage is changed outside FSDP, this use case "
```

Comment:
nit: I guess it can be changed outside of FSDP, but before first forward, i.e. we have the following in tests

model = FSDP(model)
model.half()
model(inputs)

but the issue is if its changed afeter the first iteration. Should we be more specific in this error message?

Reply:
sounds good